python/cuspatial/cuspatial/core/binpreds/binpreds.py

```diff
@@ -0,0 +1,330 @@
+        ):
+            partial_result = result.groupby("idx").sum()
+            df_result = (partial_result > 0) & (
+                partial_result < len(point_result)
+            )
+        else:
+            df_result = result.groupby("idx").sum() > 1
+        point_result = cudf.Series(
+            df_result["pip"], index=cudf.RangeIndex(0, len(df_result))
+        )
+        point_result.name = None
+        return point_result
+
+
+class IntersectsBinpred(ContainsProperlyBinpred):
+    def preprocess(self, lhs, rhs):
+        if contains_only_polygons(rhs):
+            (lhs, rhs) = (rhs, lhs)
+        return super().preprocess(lhs, rhs)
+
+
+class WithinBinpred(ContainsProperlyBinpred):
+    def preprocess(self, lhs, rhs):
+        if contains_only_polygons(rhs):
+            (lhs, rhs) = (rhs, lhs)
+        return super().preprocess(lhs, rhs)
+
+    def postprocess(self, point_indices, point_result):
+        """Postprocess the output GeoSeries to ensure that they are of the
+        correct type for the operation."""
+        result = cudf.DataFrame({"idx": point_indices, "pip": point_result})
+        df_result = result
+        # Discrete math recombination
+        if (
+            contains_only_linestrings(self.rhs)
+            or contains_only_polygons(self.rhs)
+            or contains_only_multipoints(self.rhs)
+        ):
+            # process for completed linestrings, polygons, and multipoints.
+            # Not necessary for points.
+            df_result = (
+                result.groupby("idx").sum().sort_index()
+                == result.groupby("idx").count().sort_index()
+            )
+        point_result = cudf.Series(
+            df_result["pip"], index=cudf.RangeIndex(0, len(df_result))
+        )
+        point_result.name = None
+        return point_result
+
+
+class EqualsBinpred(BinaryPredicate):
+    def preprocess(self, lhs, rhs):
+        return (lhs, rhs, None)
+
+    def postprocess(self, point_indices, point_result):
+        if isinstance(point_result, bool):
+            return point_result
+        elif len(point_result) == 1:
+            return point_result[0]
+
+    def _op(self, lhs, rhs):
+        """Compute the equals relationship between two GeoSeries."""
+        result = False
+        if contains_only_points(lhs):
+            result = lhs.points.xy.equals(rhs.points.xy)
+        elif contains_only_linestrings(lhs):
+            result = lhs.lines.xy.equals(rhs.lines.xy)
+        elif contains_only_polygons(lhs):
+            result = lhs.polygons.xy.equals(rhs.polygons.xy)
+        elif contains_only_multipoints(lhs):
+            result = lhs.multipoints.xy.equals(rhs.multipoints.xy)
+        return result
+
+
+class CrossesBinpred(EqualsBinpred):
+    def postprocess(self, point_indices, point_result):
+        if has_same_geometry(self.lhs, self.rhs) and contains_only_points(
+            self.lhs
+        ):
+            return cudf.Series([False] * len(self.lhs))
```

Comment:
Any chance we can avoid computing `_op` and early return this?

Reply:
Hmm good idea I'll consider that in the abstraction design. My first impulse is to add a bool to the `BinaryPredicate` class called `done`. `_op` and `postprocess` will only call if `not done`.

My second impulse is to not perform this calculation in `postprocess` but just do it in `_op` instead.

I will probably do another pass of the inheritance design in this PR.

Reply:
I'm also considering duck typing `_op` and replacing it with the correct predicate function if `preprocess` deigns. Alternatively I could allow specification of an alternate `_op` that would be used if defined.